### PR TITLE
fullgradients explainer added

### PR DESCRIPTION
**Pull request by kamilrzechowski**
## Description

The PR introduces the [FullGradients](https://arxiv.org/abs/1905.00780) XAI algorithm.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] the algorithm passes test_explainers.py tests and all other existing tests as well

## Checklist:

- [x] I have updated all related `pyproject.toml` files.
- [x] I have updated lock files.
- [x] I have set an Assignee and Reviewers.
- [x] I have labeled the PR correctly.
- [x] I have performed a self-review of my own code.
- [x] I have covered my code and changes with unit tests.
- [x] I have annotated types as extensively as possible.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] All code references and dependencies will work.

### CODE REVIEW

File: foxai/explainer/fullgradeints.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
+"""File with GradCAM algorithm explainer classes.
+
+Paper: https://arxiv.org/abs/1905.00780
+Based on https://github.com/idiap/fullgrad-saliency
+
+Note:
+FullGradient does not support inplace operations.
+ReLU in place operations are turned of in the model automatically.
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:8]:
```suggestion
ReLU inplace operations are turned off in the model automatically.
```

File: foxai/explainer/fullgradeints.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+Howerer inplace addition, multiplication etc. is the model design
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:9]:
```suggestion
However inplace addition, multiplication etc. is the model design
```

File: foxai/explainer/fullgradeints.py [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+choice and shlould be avoided in order to make the FullyGradient explainer work
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:10]:
```suggestion
choice and should be avoided in order to make the FullyGradient explainer work
```

File: foxai/explainer/fullgradeints.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+from foxai.array_utils import validate_result
+from foxai.explainer.base_explainer import CVExplainer
+from foxai.logger import create_logger
+
+_LOGGER: Optional[logging.Logger] = None
+
+_INPLACE_ERROR_MSG: Final[
+    str
+] = """
+Error possibly coused by in place operations.
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:44]:
```suggestion
Error possibly caused by inplace operations.
```

File: foxai/explainer/fullgradeints.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+FullGradient does not support inplace operations.
+ReLU in place operations are turned of in the model automatically.
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:46]:
```suggestion
ReLU inplace operations are turned off in the model automatically.
```

File: foxai/explainer/fullgradeints.py [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+Howerer inplace addition, multiplication etc. is the model design
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:47]:
```suggestion
However inplace addition, multiplication etc. is the model design
```

File: foxai/explainer/fullgradeints.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+choice and shlould be avoided in order to make the FullyGradient explainer work
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:48]:
```suggestion
choice and should be avoided in order to make the FullyGradient explainer work
```

File: foxai/explainer/fullgradeints.py [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+    global _LOGGER
+    if _LOGGER is None:
+        _LOGGER = create_logger(__name__)
+    return _LOGGER
+
+
+class FullGradExtractor:
+    """Extract gradient tensors needed for FullGrad computation using pytorch hooks."""
+
+    def __init__(self, model, im_size: Tuple[int, int, int] = (3, 224, 224)):
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:73]:
```suggestion
    def __init__(self, model: torch.nn.Module, im_size: Tuple[int, int, int] = (3, 224, 224)):
```

File: foxai/explainer/fullgradeints.py [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+                    torch.nn.SELU,
+                ),
+            ):
+                log().warning(
+                    "The FullGradients algorithm only works properly with ReLu and "
+                    "linear activation functions by design."
+                    f", however {module} layer found in model definition."
+                )
+
+    def _extract_layer_bias(self, module) -> Optional[torch.Tensor]:
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:115]:
```suggestion
    def _extract_layer_bias(self, module: torch.nn.Module) -> Optional[torch.Tensor]:
```

File: foxai/explainer/fullgradeints.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        else:
+            return module.bias.data
+
+    @property
+    def biases(self) -> List[torch.Tensor]:
+        """Get biases."""
+        return self._biases
+
+    def _extract_layer_grads(  # pylint: disable = unused-argument
+        self, module, in_grad: _grad_t, out_grad: _grad_t
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:154]:
```suggestion
        self, module: torch.nn.Module, in_grad: _grad_t, out_grad: _grad_t
```

File: foxai/explainer/fullgradeints.py [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+    ) -> None:
+        """Collect gradient outputs from each layer, that contains bias term.
+
+        Args:
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:158]: Missing `module` argument.

File: foxai/explainer/fullgradeints.py [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+
+        return input_gradients, self.feature_grads
+
+
+class FullGrad:
+    """
+    Compute FullGrad saliency map and full gradient decomposition
+    """
+
+    def __init__(self, model, image_size=(3, 224, 224)):
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:195]:
```suggestion
    def __init__(self, model: torch.nn.Module, image_size=(3, 224, 224)):
```

File: foxai/explainer/fullgradeints.py [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+
+        cuda = next(self.model.parameters()).is_cuda
+        self._device = torch.device("cuda" if cuda else "cpu")
+        _ = self.check_completeness()
+
+    def check_completeness(self) -> bool:
+        """Check if completeness property is satisfied. If not, it usually means that
+        some bias gradients are not computed (e.g.: implicit biases of non-linearities).
+
+        Returns:
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:211]: Add a description of raised exceptions.

File: foxai/explainer/fullgradeints.py [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+            whether the completnes condition is fulfielled
+        """
+
+        # Random input image
+        sample_input: torch.Tensor = torch.randn(self.im_size).to(self._device)
+
+        # Get raw outputs
+        self.model.eval()
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:219]: In this line, we are modifying the model's state. We should restore somehow restore it at the end.

File: foxai/explainer/fullgradeints.py [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+        self.model: torch.nn.Module = model
+        self.im_size: Tuple[int, int, int, int] = (1,) + image_size
+        self.model_ext: FullGradExtractor = FullGradExtractor(
+            model=self.model, im_size=image_size
+        )
+        self._biases: List[torch.Tensor] = self.model_ext.biases
+
+        cuda = next(self.model.parameters()).is_cuda
+        self._device = torch.device("cuda" if cuda else "cpu")
+        _ = self.check_completeness()
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:205]: `check_completeness` returns boolean which is discarded. We should use the result of this function or remove returning value.

File: foxai/explainer/fullgradeints.py [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+        Args:
+            image: the input image to the nn
+            target_class: the class for which to draw explanations. If None, the argmax of output is chosen.
+
+        Returns:
+            input image gradients (contribution of each input pixel to the final output)
+            list of contributions of bias terms in each layer in the model (each layer with the bias term)
+        """
+
+        self.model.eval()
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:264]: Same as above, we should somehow revert this change to the model.

File: foxai/explainer/fullgradeints.py [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+            input_data: the input image to the model
+            target: the target class to be explained. If none the argmax class is chosen.
+
+        Returns:
+            silency map
+        """
+        if isinstance(target, int):
+            target = torch.tensor(data=[target] * input_data.shape[0]).to(self._device)
+
+        self.model.eval()
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:343]: Same as above, we should somehow revert this change to the model.

File: foxai/explainer/fullgradeints.py [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+                tensor as input, a single input tensor should be provided.
+            pred_label_idx: Output indices for
+                which gradients are computed (for classification cases,
+                this is usually the target class).
+                If None provided, the argmax of output is chosen for explanation.
+
+                Default: None
+
+        Returns:
+            Element-wise product of (upsampled) GradCAM
```
Comment (adamwawrzynski) [inline foxai/explainer/fullgradeints.py:411]: I think something is wrong here - mentioned GradCAM and Guided Backprop and not FullGrad. Can you take a look at this?

**Review by adamwawrzynski — APPROVED** [review]
Overall, LGTM.

**Review by rpytel1 — APPROVED** [review]

File: foxai/explainer/fullgradeints.py (hunk 15 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: fixed

File: foxai/explainer/fullgradeints.py (hunk 17 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done

File: foxai/explainer/fullgradeints.py (hunk 16 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done

File: foxai/explainer/fullgradeints.py (hunk 14 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done

File: foxai/explainer/fullgradeints.py (hunk 13 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py:235]: done

File: foxai/explainer/fullgradeints.py (hunk 10 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done

File: foxai/explainer/fullgradeints.py (hunk 12 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done

File: foxai/explainer/fullgradeints.py (hunk 11 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py:168]: done

File: foxai/explainer/fullgradeints.py (hunk 8 above)
Comment (kamilrzechowski) [reply foxai/explainer/fullgradeints.py]: done
